components/server/src/server.rs

```diff
@@ -1851,3 +1854,118 @@ impl IOBudgetAdjustor for EnginesResourceInfo {
         (total_budgets as f32 * score) as usize
     }
 }
+
+// currently, only support return json result
+pub fn get_flatten_cfg_info(cfg: &TiKvConfig) -> Vec<Value> {
+    fn append_sub_key(key: &mut String, sub_key: &str) {
+        if !key.is_empty() {
+            key.push('.');
+        }
+
+        key.push_str(sub_key);
+    }
+
+    fn pop_sub_key(key: &mut String, sub_key: &str) {
+        let len = if key.len() > sub_key.len() {
+            key.len() - sub_key.len() - 1
+        } else {
+            0
+        };
+        key.truncate(len);
+    }
+
+    fn to_cfg_value(default_value: &Value, cfg_value: Option<&Value>, key: &str) -> Value {
+        let mut res = Map::with_capacity(2);
+        res.insert("Name".into(), Value::String(key.into()));
+        res.insert("DefaultValue".into(), default_value.clone());
+        if let Some(cfg_val) = cfg_value {
+            if default_value != cfg_val {
+                res.insert("ValueInFile".into(), cfg_val.clone());
+            }
+        }
+
+        Value::Object(res)
+    }
+
+    // configs that should not be flatten because the config type is HashMap instead of submodule.
+    lazy_static! {
+        static ref NO_FLATTEN_CFGS: HashSet<&'static str> = {
+            let mut set = HashSet::new();
+            set.insert("server.labels");
+            set
+        };
+    }
+
+    fn flatten_value(
+        default_obj: &Map<String, Value>,
+        value_obj: &Map<String, Value>,
+        key_buf: &mut String,
+        res: &mut Vec<Value>,
+    ) {
+        for (k, v) in default_obj.iter() {
+            let cfg_val = value_obj.get(k);
+            append_sub_key(key_buf, k);
```

Comment:
These tiny functions don't make it easy to review. How about
```suggestion
            let old_len = key_buf.len();
            if !k.is_empty() {
              key.push_str(k);
              key.push('.');
            }
```